### ISSUE TITLE
Request: deep_copy within parallel regions

### ISSUE THREAD

**Issue by vbrunini** [kokkos/kokkos#689]
I'd like to have versions of deep_copy (both the copy from another view and the fill with a constant value) that can be called from within parallel regions and are as performant as the equivalent using memcpy/memset on the underlying data.

@crtrott

**Comment by gmackey** [comment]
This functionality is something that could be supported, but I'm not sure using the function deep_copy() is the way we want to go.  What you want is something that copies between views only in the same memory space.  deep_copy() implies that it works copying between memory spaces.

@crtrott We've talked a couple of times about adding some of the stl algorithm functionality to Kokkos algorithms.  One way to approach this problem is to use a Kokkos algorithms analogue to stl::copy().  This would avoid confusing the intent of the name deep_copy().  Although it would bring up the issue of should our copy() work for both in parallel regions and outside parallel regions?  Should we have separate functions that do each (copy() and serial_copy()), or should a single function detect if it is in a parallel region and do the right thing?

**Comment by mhoemmen** [comment]
deep_copy is the "easiest Kokkos kernel" ;-)  Would it make sense to have team-level, thread-level, etc. versions of it?

**Comment by hcedwar** [comment]
@vbrunini : what is the scope you are requesting?  As per @mhoemmen comment?  Restricted to same memory space?

**Comment by vbrunini** [comment]
Restricted to the same memory space and just single thread level is fine for my needs.

**Comment by ibaned** [comment]
Is this still needed, and if so is it okay if its called something different (like `local_deep_copy`)?

**Comment by vbrunini** [comment]
Yes this would still be useful. I'm fine with a different name, maybe fill to be consistent with std::fill?

**Comment by Char-Aznable** [comment]
I have a similar request for this feature. Here's a usage example:
```
#include <cstdio>
#include <iostream>
#include <Kokkos_Core.hpp>
#include <Kokkos_DualView.hpp>

using namespace std;

using ExecutionSpace = Kokkos::DefaultExecutionSpace;

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv); {
  Kokkos::DualView<int***[5][6],Kokkos::LayoutRight,ExecutionSpace,Kokkos::MemoryTraits<Kokkos::Atomic>> a("a",2,3,4);
  using MemorySpace = decltype(a)::memory_space;
  using HostMirrorSpace = decltype(a)::host_mirror_space;
  auto asub = Kokkos::subview(a,0,1,2,Kokkos::ALL(),Kokkos::ALL());
  cout << "Before: \n";
  for(int i = 0; i < 5; ++i) {
    for(int j = 0; j < 6; ++j) {
      asub.h_view(i,j) = i * j;
      cout.width(5);
      cout << asub.h_view(i,j) << " ";
    }
    cout << endl;
  }
  a.modify<HostMirrorSpace>();
  a.sync<MemorySpace>();
  Kokkos::parallel_for(Kokkos::RangePolicy<ExecutionSpace>(0,5), KOKKOS_LAMBDA (const int i){
    auto subSrc = Kokkos::subview(a.d_view,0,1,2,Kokkos::ALL(),i+1);
    auto subDst = Kokkos::subview(a.d_view,0,1,2,Kokkos::ALL(),i);
    Kokkos::deep_copy(subDst,subSrc);
    }); 
  Kokkos::fence();
  a.modify<MemorySpace>();
  a.sync<HostMirrorSpace>();

  cout << "After: \n";
  for(int i = 0; i < 5; ++i) {
    for(int j = 0; j < 6; ++j) {
      cout.width(5);
      cout << asub.h_view(i,j) << " ";
    }
    cout << endl;
  }

  } Kokkos::finalize();
}
```
This example is a bit abusive because it basically copy the `i+1`th row to the `i`th row of a subview matrix but the actually use case would be `i` being somewhat randomly associated with each thread.

**Comment by mhoemmen** [comment]
@vbrunini I was thinking about how to do this.  The main issue is that `Kokkos::TeamPolicy` doesn't expose `member_type`, so we can't overload `Kokkos::deep_copy` on it.  TeamPolicy's base class _does_ expose `member_type`, but the base class lives in the `Kokkos::Impl` namespace, and is thus not part of the public interface.

**Comment by vbrunini** [comment]
I don't think a solution involving ```Kokkos::TeamPolicy``` or its ```member_type``` makes sense because we may also want to call this from loops that use a RangePolicy.

**Comment by mhoemmen** [comment]
@vbrunini Why couldn't we just call `Kokkos::deep_copy (Kokkos::Serial, dst, src);` in that case?

**Comment by vbrunini** [comment]
I hadn't thought about that option, no objections to that.

**Comment by ibaned** [comment]
@mhoemmen I think we've seen issues before when trying to run a `Serial` `parallel_for` inside some other `parallel_for`. There is an interpretation in which that should work, however in practice I think `Serial` is an outer-level execution space only.

**Comment by mhoemmen** [comment]
@ibaned The `member_type` of a RangePolicy is an integer; we could just overload `Kokkos::deep_copy` so if the first argument is an integer, it does a sequential copy.  That may look weird, though.
